### ISSUE TITLE
Create clickable links for dimensions

### ISSUE THREAD

**Issue by owlas** [lightdash/lightdash#2470]
### Description

If I have a dimension with a Hubspot id or a GitHub pull request, I'd like to be able to click the dimension value and be linked to an external website for that resource!

E.g. something like this in the .yml files:

```
columns:
  name: screen_id
    meta:
      dimension: 
        url: "github.com/lightdash/screens/${screen_id}"
```

- you can add constants for the URL
- OR you can add some variable for the URL (e.g. here, I added the `screen_id` to the end of the URL)

**Comment by Martin-Carlsson** [comment]
@owlas 
This is something I'm asked about.

I think a combination of markdown link would make it stronger, like:

```
columns:
  name: screen_id
    meta:
      dimension: 
        url: "[Click screen: ${screen_id}](https://github.com/lightdash/screens/${screen_id})"
```

Another implementation could be:
```
columns:
  name: screen_link
    meta:
      dimension: 
        format: url
```
With the value of `screen_link` being: `[Click screen: 1234}](https://github.com/lightdash/screens/1234)`

With the use `format`, it would also be possible to add a clickable link as a table calculation. _(if you add the option of choosing format in table calculations)_

**Comment by TuringLovesDeathMetal** [comment]
> `url: "[Click screen: ${screen_id}](https://github.com/lightdash/screens/${screen_id})"`

This is a neat implementation @Martin-Carlsson 

We already let you change the dimension label, so we could maybe do something like this instead:

```
columns:
  name: my_cool_dim
    meta:
      dimension:
        label: 'Click here: ${my_cool_dim}'
        url: www.katiesurl.com
```

Otherwise, where were you expecting the URL markdown text to appear? For example, if I had a table and I added `my_cool_dim` to it, would the result table have the dimension values? Or the URL markdown label values?

And if you added a `dimension label`, would the URL markdown override that label?

**Comment by ZeRego** [comment]
@TuringLovesDeathMetal From what I understand we want to have the link in the table cell. I think this could be a format configuration similar to how we can format numbers. 

Like Martin suggested, we could support markdown which would be more powerful. 
```
columns:
    - name: screen_id
      meta:
        dimension:
          format: 'markdown'
          markdownTemplate:  "[Click screen: ${screen_id}](https://github.com/lightdash/screens/${screen_id})"
```

Implementation:
- The markdown template should have access to the entire row data. So we can build a url that needs the user id and the project id. 
- Should the template support logical operations ? I don't think we should for now as this can bring security concerns. For now I suggest just having references.
- Should this formatting work when used in charts ? Not sure how easy it would be for echarts to support this.

**Comment by TuringLovesDeathMetal** [comment]
riiiiight - sorry, I hadn't had my coffee yet when I replied 🙃 

Okay, makes sense. Ya, it could be cool to allow for markdown then allow a template. 

> Should this formatting work when used in charts ?

Only in Table + Big Value visualizations. 

> Should the template support logical operations ?

Not sure I understand what you mean by this... can you add logical operations in Markdown? 🤔

**Comment by ZeRego** [comment]
> Not sure I understand what you mean by this... can you add logical operations in Markdown? 🤔

This is a markdown + templating ( so we can inject values into it, like Jinja ) 

If we allowed logical operations the user could have something like `if user id is 1, show a different link`. That means we would have to compute the code and potential malicious code injection. Some templating libraries are more secure than others about how they compile/execute the code.

**Comment by TuringLovesDeathMetal** [comment]
ah, makes sense. 

Ya, let's not allow that for now

**Comment by Martin-Carlsson** [comment]
> **The markdown template should have access to the entire row data.** 

🙏👌🤩

> **Should the template support logical operations**

I could see the value down the road. But for **our** use case it is not a priority.

_I am **not** a skilled software developer, and I do **not** have the full overview of Lightdash. So take what I say with a huge grain of salt._ My gut feeling is that decisions like this, at this stage of development, might create issues later in development.

> **Only in Table + Big Value visualizations**

🙏👌🤩

>  #2472 

Yea, that could be a cool implementation

**Comment by ZeRego** [comment]
> I could see the value down the road. But for our use case it is not a priority.
> I am not a skilled software developer, and I do not have the full overview of Lightdash. So take what I say with a huge grain of salt. My gut feeling is that decisions like this, at this stage of development, might create issues later in development.

@Martin-Carlsson  The user can always use SQL to achieve such advanced use cases. 

E.g:
```
- name: custom_url
   meta:
     dimension:
        type: string
        sql: IF(user_id<1000, "https://example.com", "https://another-example.com");
        format: 'markdown'
        markdownTemplate:  "[Click my custom url](${custom_url})"
```

**Comment by ZeRego** [comment]
@all-contributors add @Martin-Carlsson for ideas

**Comment by owlas** [comment]
This would couple together formatting + url. 

Can we imagine a case where the user would want to inject the formatted value into the url?

e.g. if I have this dimension

```yaml
columns:
  - name: salary
    meta: 
      dimension:
        format: usd
        url: 'https://google.com/search?q=${ value } in gbp'
```

^ weird example that would let you click the dollar sign and see it in uk pounds via google. But should the `$` get included in the value? 

If we keep the format and url separate we could expose:
 - `${ raw_value }` - would give `1000000`
 - `${ value }` - would give `$1,000,000.00`


If we keep the formatting and and url together (using markdown format) then the user would have to do something like:

```
url: 'https://google.com/search?q=\$${ value }.00 in gbp'
```


Sorry for the terrible example, it's getting late here!! But the point is, do we want adding a url to eject the user from our formatting options?

We could actually have both and introduce markdown formatting as an additional feature.


---


Completely separate concern, do we want multiple links with names?

e.g. I click a company name and it gives me the option to:
* open in Hubspot
* search in google
* view details in clearbit

If so we'd have something like:

```yaml
columns:
  - name: company_name
    meta: 
      dimension:
        format: usd
        urls:
          - url: 'https://hubspot.com....'
            label: 'Show in Hubspot!'
          - url: 'https://clearbit.com...'
            label: 'View details in clearbit'
```

I guess these would have to go on the context menu with "show underlying values" and "filter by" if there are many options.

---

Thinking about these two together makes me realise they are perhaps two separate features:

1. "markdown formatting" - gives templated markdown option for more formatting controls including to actually render the dimension/metric as a clickable hyperlink
2. "url actions" - adds actions via urls to the context menu along with "show underlying values"

wdyt?

**Comment by owlas** [comment]
^ just one more consideration, having the formatting as markdown would also restrict us in plain text environments (like slack or other integrations, where the value of the query has markdown specific formatting)

**Comment by TuringLovesDeathMetal** [comment]
1. Not sure I understand what's happening with the markdown/formatting stuff. I'm a bit confused what we're trying to do (or avoid doing) here 🤔 
2. Love the idea of having multiple URLs! Once Irakli builds the data interactivity stuff this week with hyperlinks, then we'll have a good idea of how this would fit in with teh existing design 😍 
3. Splitting this out into "URL actions" and "markdown formatting" makes sense to me, I think? I'm just not sure when I'd use `markdown formatting`. It feels like we're taking all potential formatting options and saying "you can just do this all in one go using fully custom formatting". Which is something we've actually talked about doing anyways!

**Comment by owlas** [comment]
Actually just realised, this work in progress: https://github.com/lightdash/lightdash/issues/2785 @PriPatel the design actually is for the urls to appear in the context menu? not on click?

**Comment by PriPatel** [comment]
@owlas yes, I changed the spec yesterday as Irakli's been progressing with the experiments. We will have conflicts in situations like this:  
![image](https://user-images.githubusercontent.com/12776522/185084657-1e2e5827-de94-448c-89ed-8b91b6614369.png)

**Comment by owlas** [comment]
Awesome! So if it's in the context menu, let's go with the multiple urls approach? Where it doesn't actually change the formatting of the dimension but will add some actions you can take from the context menu.

We'll also have to add that along with the other context menus on charts etc.

**Comment by owlas** [comment]
Moving to blocked until the table context menu has been updated: https://github.com/lightdash/lightdash/issues/2785

**Comment by TuringLovesDeathMetal** [comment]
@owlas i think this is unblocked now!

**Comment by owlas** [comment]
Yep!
